src/instructlab/training/logger.py

```diff
@@ -0,0 +1,177 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Standard
+from collections.abc import Mapping
+from datetime import datetime
+from pathlib import Path
+import json
+import os
+
+try:
+    # Third Party
+    import wandb
+except ImportError:
+    wandb = None
+
+try:
+    # Third Party
+    from torch.utils.tensorboard import SummaryWriter
+except ImportError:
+    SummaryWriter = None
+
+# First Party
+from instructlab.training import async_logger
+
+
+def flatten_dict(
+    d: Mapping,
+    sep: str = ".",
+    prefix: str = "",
+    _flattened: dict | None = None,
+):
+    if _flattened is None:
+        _flattened = {}
+
+    for k, v in d.items():
+        if isinstance(v, Mapping):
+            flatten_dict(v, sep=sep, prefix=f"{prefix}{k}{sep}", _flattened=_flattened)
+        else:
+            _flattened[prefix + k] = v
+
+    return _flattened
+
+
+class BaseLogger:
+    def __init__(self, run_name: str | None, log_dir: str | Path):
+        if run_name is None:
+            # default template
+            run_name = "{time}_rank{rank}"
+
+        substitutions = {
+            "{time}": datetime.now().isoformat(),
+            "{rank}": os.environ.get("RANK", 0),
+            "{local_rank}": os.environ.get("LOCAL_RANK", 0),
+        }
+        for placeholder_pat, value in substitutions.items():
+            run_name = run_name.replace(placeholder_pat, str(value))
+
+        self.run_name = run_name
+        self.log_dir = log_dir
+
+    def setup(self):
+        pass
+
+    def log_dict(self, d: dict, step: int | None = None):
+        pass
+
+    def teardown(self, exit_code: int = 0):
+        pass
+
+
+class FileLogger(BaseLogger):
+    def setup(self):
+        filename = Path(self.log_dir) / f"{self.run_name}.jsonl"
+        os.makedirs(filename.parent, exist_ok=True)
+
+        self.file = open(filename, "a")  # pylint: disable=R1732
+
+    def log_dict(self, d: dict, step: int | None = None):
+        if step is not None and not "step" in d:
+            d["step"] = step
+
+        self.file.write(json.dumps(d, indent=None) + "\n")
+
+    def teardown(self, exit_code: int = 0):
```

Comment:
Better to indicate these are unused

```suggestion
    def teardown(self, **unused_kwargs):
```